core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1731,8 +1751,14 @@ private[spark] class DAGScheduler(
             // resubmitted stage attempt.
             mapOutputTracker.unregisterAllMapOutput(shuffleId)
           } else if (mapIndex != -1) {
-            // Mark the map whose fetch failed as broken in the map stage
-            mapOutputTracker.unregisterMapOutput(shuffleId, mapIndex, bmAddress)
```

Comment:
Probably, comparing the `mapId` of the MapStatus should be enough for this case. But it should only work when we use the new shuffle fetch protocol (enabled by default). For example,


```scala
val mapStatus = mapOutputTracker.getMapOutput(shuffleId, mapIndex)
if (mapStatus.mapId == fetchFailed.mapId) {
  mapOutputTracker.unregisterMapOutput(shuffleId, mapIndex, bmAddress)
}
```

For the old protocol, I think we can at least compare `bmAddress` to prevent most cases. Although, it still cannot fix the case where `bmAddress` is the same.